File: java/binarysearchtree/BinarySearchTree.java
class BinarySearchTree {

    private Node root = null;

    public BinarySearchTree(Node root) {
        this.root = root;
    }
    
    public void getInorder(Node node) {
        if (node != null) {
            getInorder(node.left);
            System.out.println(node.getValue());
            getInorder(node.right);
        }
    }

    public void getPreorder(Node node) {
        if (node != null) {
            System.out.println(node.getValue());
            getPreorder(node.left);
            getPreorder(node.right);
        }        
    }

    public void getPostorder(Node node) {
        if (node != null) {
            getPostorder(node.left);
            getPostorder(node.right);
            System.out.println(node.getValue());
        }        
    }
    
    public void insert(int value) {
        if (root == null) {
            root = new Node(value, null, null);
        } else {
            root.insertChild(value);
        }
    }

    public static void main(String[] args) {
        BinarySearchTree bst = new BinarySearchTree(null);
        bst.insert(5);
        bst.insert(1);
        bst.insert(7);
        bst.getInorder(bst.root);
    }
}

File: python/datastructures/binarysearchtree/binary_search_tree.py
#!/usr/bin/env python3

from node import Node

"""
Binary search tree implementation
"""

class BinarySearchTree(object):

    def __init__(self, root=None):
        self.root = root

    def get_inorder(self, node, values):
        if node is not None:
            self.get_inorder(node.left, values)
            values.append(node.value)
            self.get_inorder(node.right, values)

    def get_preorder(self, node, values):
        if node is not None:
            values.append(node.value)
            self.get_preorder(node.left, values)
            self.get_preorder(node.right, values)

    def get_postorder(self, node, values):
        if node is not None:
            self.get_postorder(node.left, values)
            self.get_postorder(node.right, values)
            values.append(node.value)

    def values(self, order="inorder"):
        values = [] # Save values in list to print out
        if order == "inorder":
            self.get_inorder(self.root, values)
        elif order == "preorder":
            self.get_preorder(self.root, values)
        else:
            self.get_postorder(self.root, values)
        return values
    
    def insert(self, value):
        if self.root is None:
            self.root = Node(value)
        else:
            self.root.insert_child(value)

# Testing
bst = BinarySearchTree()
bst.insert(5)
bst.insert(1)
bst.insert(7)
print(bst.values())
